WIP: Add video instance segmentation support for mask2former

shivalikasingh95 opened the pull request:
```
import torch
from PIL import Image
from transformers import Mask2FormerImageProcessor, Mask2FormerForUniversalSegmentation, Mask2FormerConfig
import torchvision

VIDEO_FILE_NAME = "/workspace/transformers/sample.mp4"

config = Mask2FormerConfig()
processor = Mask2FormerImageProcessor(config)
model = Mask2FormerForUniversalSegmentation(config)

video_tensor = torchvision.io.read_video(VIDEO_FILE_NAME)[0]

processed_inputs = []
for frame in video_tensor:
    inputs = processor(images=frame, return_tensors="pt", do_resize=True, size=(480, 640))
    processed_inputs.append(inputs.pixel_values)

pixel_values = torch.cat(processed_inputs)

model.eval()
with torch.no_grad():
    outputs = model(pixel_values=pixel_values, video_input=True)
```


Hi @alaradirik this is how I'm currently preparing video input and feeding to the model.
Note that I'm using a `video_input` flag while passing the input to the model to tell the model what kind of input has been passed. 
I felt it would be better to do it this way rather than via config file. But let me know if you feel otherwise. 

Last time we discussed about having some kind of video_utils file. But I'm not sure how to go about designing that. If you can guide me regarding it then it would be great or we can collaborate on it too.

shivalikasingh95 committed:
video mask2former initial commit

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former-video

shivalikasingh95 committed:
minor fixes

shivalikasingh95 committed:
Merge branch 'add-mask2former-video' of https://github.com/shivalikasingh95/transformers into add-mask2former-video

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -103,6 +103,14 @@ class Mask2FormerConfig(PretrainedConfig):
             Feature strides corresponding to features generated from backbone network.
         output_auxiliary_logits (`bool`, *optional*):
             Should the model output its `auxiliary_logits` or not.
+        num_frames (`int`, *optional*, defaults to 2):
```
Could you also add a `is_video` or `video_input` argument to differentiate between image and video segmentation models?

shivalikasingh95 replied:
sure

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -103,6 +103,14 @@ class Mask2FormerConfig(PretrainedConfig):
+            Range of frames in the video
+        frame_shuffle (`bool`, *optional*, defaults to False): 
+            Whether to shuffle the frames or not
+        input_augmentations(`List[str]`, *optional*, defaults to []):
+            List of augmentations to apply on the video input such as brightness", "contrast", etc.
```
I believe the input augmentations are only applied during training. We are currently not supporting any data augmentations within the `ImageProcessor` classes and encourage users to combine external libraries (e.g. albumentations) with our preprocessor to do so. 

```suggestion
```

shivalikasingh95 replied:
ah right, sorry I forgot!

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -154,6 +162,10 @@ def __init__(
+        num_frames: int = 2,
+        frame_range: int = 20,
+        frame_shuffle: bool = False,
+        input_augmentations: List[str] = [],
```
Please see my comment above, we can get rid of this.
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -843,6 +870,51 @@ def multi_scale_deformable_attention(
     )
     return output.transpose(1, 2).contiguous()
 
```
Missing a space over here :)
```suggestion


```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -843,6 +870,51 @@ def multi_scale_deformable_attention(
+        self.scale = 2 * math.pi if scale is None else scale
+
+    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
+        
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -843,6 +870,51 @@ def multi_scale_deformable_attention(
+    """
+
+    def __init__(
+        self, num_pos_feats: int = 64, temperature: int = 10000, normalize: bool = False, scale: Optional[float] = None
```
Could any of these be included in the config file instead?

shivalikasingh95 replied:
sure, I can add these.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -843,6 +870,51 @@ def multi_scale_deformable_attention(
+        self.normalize = normalize
+        self.scale = 2 * math.pi if scale is None else scale
+
+    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
```
It's preferable to use more descriptive argument names:
```suggestion
    def forward(self,  hidden_states: Tensor, mask: Optional[Tensor] = None) -> Tensor:
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -843,6 +870,51 @@ def multi_scale_deformable_attention(
+    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
+        
+        if mask is None:
+            mask = torch.zeros((x.size(0), x.size(2), x.size(3), x.size(4)), device=x.device, dtype=torch.bool)
```
```suggestion
            mask = torch.zeros(hidden_states.shape[:4]), device=hidden_states.device, dtype=torch.bool)
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2224,6 +2391,7 @@ def forward(
         self,
         pixel_values: Tensor,
         pixel_mask: Optional[Tensor] = None,
+        video_input: Optional[bool] = False,
```
Adding new arguments to the forward functions of existing models create backward compatibility issues, could you add this as a configuration parameter instead?
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2272,12 +2440,20 @@ def forward(
-            output_hidden_states=True,
-            output_attentions=output_attentions,
-        )
+        if video_input:
```
```suggestion
        if self.config.video_input:
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2215,7 +2382,7 @@ def __init__(self, config: Mask2FormerConfig):
         super().__init__(config)
         self.pixel_level_module = Mask2FormerPixelLevelModule(config)
         self.transformer_module = Mask2FormerTransformerModule(in_features=config.feature_size, config=config)
-
+        self.video_transformer_module = Mask2FormerVideoTransformerModule(in_features=config.feature_size, config=config)
```
It'd be preferable to not modules unless they will be used during inference. Could you convert this to an if-else statement instead?

```
        if config.video_input:
             self.transformer_module = Mask2FormerTransformerModule(in_features=config.feature_size, config=config)
        else:
             self.transformer_module = Mask2FormerVideoTransformerModule(in_features=config.feature_size, config=config)
```

shivalikasingh95 replied:
aah yes, I get my mistake. I'll fix this.

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2272,12 +2440,20 @@ def forward(
-            output_attentions=output_attentions,
-        )
+        if video_input:
+            transformer_module_output = self.video_transformer_module(
```
```suggestion
            transformer_module_output = self.transformer_module(
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2373,6 +2551,7 @@ def forward(
         mask_labels: Optional[List[Tensor]] = None,
         class_labels: Optional[List[Tensor]] = None,
         pixel_mask: Optional[Tensor] = None,
+        video_input: Optional[bool] = False,
```
Please see my comment above, it's better to remove this to avoid backward compatibility issues
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2384,7 +2563,8 @@ def forward(
             list of target class labels of shape `(num_labels, height, width)` to be fed to a model. They identify the
             labels of `mask_labels`, e.g. the label of `mask_labels[i][j]` if `class_labels[i][j]`.
-
+        video_input (`bool`, *optional`):
+            Whether input given to the model is a video.
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/modeling_mask2former.py:
```diff
@@ -2496,10 +2676,11 @@ def forward(
         outputs = self.model(
             pixel_values=pixel_values,
             pixel_mask=pixel_mask,
+            video_input=video_input,
```
```suggestion
```

alaradirik commented on src/transformers/models/mask2former/configuration_mask2former.py:
```diff
@@ -204,6 +216,10 @@ def __init__(
+        self.num_frames = num_frames
+        self.frame_range = frame_range
+        self.frame_shuffle = frame_shuffle
+        self.input_augmentations = input_augmentations
```
```suggestion
```

alaradirik reviewed (commented):
Thank you for adding this!

I think the PR looks great apart from the minor bugs. My main comments are suggestions are as follows:
- I think it's fine to transform the video input to a list of images separately, models such as `VideoMAE` also does this and I can do a follow up PR to create video_utils.py if needed
- is_video or video_input argument should best be included in the configuration parameters as adding it to the forward functions create a backward compatibility issues
- Could you add one or two integration tests to test_modeling_mask2former.py?

shivalikasingh95 committed:
Apply suggestions from code review

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former-video

shivalikasingh95 committed:
change video_input to is_video. Read is_video value from config.

shivalikasingh95 committed:
fix forward pass

shivalikasingh95 committed:
update model doc

shivalikasingh95 committed:
update examples to depict usage in video mode

alaradirik committed:
Add AutoModelForZeroShotImageClassification (#22087)

Adds AutoModelForZeroShotImageClassification to transformers

wdp-007 committed:
add new model of MGP-STR (#21418)

* add new model of MGP-STR

* fix the check failings

* remove torch and numpy from mgp_tokenization

* remove unused import from modeling_mgp_str

* add test_processing_mgp_str

* rm test_processing_mgp_str.py

* add test_processing_mgp_str

* add test_processing_mgp_str

* add test_processing_mgp_str

* rm test_processing_mgp_str and add softmax outs to model

* rm test_processing_mgp_str and add softmax outs to model

* rewrite the code of mgp-str according to PR suggestions

* rewrite the code of mgp-str according to PR suggestions

* add new model of MGP-STR

* fix the check failings

* remove torch and numpy from mgp_tokenization

* remove unused import from modeling_mgp_str

* add test_processing_mgp_str

* rm test_processing_mgp_str.py

* add test_processing_mgp_str

* add test_processing_mgp_str

* add test_processing_mgp_str

* rm test_processing_mgp_str and add softmax outs to model

* rewrite the code of mgp-str according to PR suggestions

* rewrite the code of mgp-str according to PR suggestions

* remove representation_size from MGPSTRConfig

* reformat configuration_mgp_str.py

* format test_processor_mgp_str.py

* add test for tokenizer and complete model/processer test and model file

* rm Unnecessary tupple in modeling_mgp_str

* reduce hidden_size/layers/label_size in test_model

* add integration tests and change MGPSTR to Mgpstr

* add test for logit values

* reformat test model file

---------

Co-authored-by: yue kun <yuekun.wp@alibaba-inc.com>

alexcalabrese committed:
Add pr_checks.mdx Italian translation (#17459) (#22116)

* Add pr_checks.mdx Italian translation (#17459)

* Updated pr_checks.mdx Italian translation (#17459)

KMFODA committed:
Fix gradient checkpointing bug in xglm (#22127)

KMFODA committed:
Fix gradient checkpointing bug in Trajectory Transformer (#22125)

KMFODA committed:
Fix gradient checkpointing bug in xlm_roberta_xl (#22128)

nickprock committed:
Added big_models.mdx italian translation #17600  (#22115)

* updated toctree

* italian translation big_model.mdx

* italian translation big_models

younesbelkada committed:
[`Blip2`] skip accelerate test (#22124)

skip accelerate test

KMFODA committed:
Fix gradient checkpointing bug in xmod (#22129)

KMFODA committed:
Fix gradient checkpointing bug in LongT5 (#22130)

KMFODA committed:
Fix gradient checkpointing bug in trocr (#22126)

* Fix gradient checkpointing bug in trocr

* Fix format

* Update src/transformers/models/trocr/modeling_trocr.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

---------

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

MKhalusova committed:
Zero-shot image classification task guide (#22132)

* WIP

* WIP

* manual inference example

* make style

* Apply suggestions from code review

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

---------

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

sgugger committed:
Fix doc link for MGP-STR (#22138)

pmollerus23 committed:
Adding Type Hints to TF_Pegasus model (#21941)

* Adding Type Hints to TF_Pegasus model

* Updated some parameters per maintainer comments

ydshieh committed:
Add a new script to check model testers' config (#22063)

* Add script

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

bishmdl76 committed:
Update configuration_align.py (projected_dim=640) (#22139)

Update configuration_align.py

updated projected_dim=640 from 512 in arguments of AlignConfig

younesbelkada committed:
[`Whiper`] add `get_input_embeddings` to `WhisperForAudioClassification` (#22133)

* add `get_input_embeddings` to `WhisperForAudioClassification`

* add common tests

* fix another common test

* Update tests/models/whisper/test_modeling_whisper.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* fix style

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

gante committed:
Trainer: let generate pick its inputs (#22108)

* Let generate pick its inputs

* fix squad seq2seq example

sgugger committed:
Enforce same behavior as PyTorch 2.0 for older versions (#22136)

stas00 committed:
[trainer] fix bug in grad accum with multiple epochs (#22098)

* [trainer] fix bug in grad accum

* comment out debug

* fix one-off

* rename counter

stas00 committed:
[deepspeed docs] Activation Checkpointing (#22099)

* [deepspeed docs] Activation Checkpointing

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update deepspeed.mdx

---------

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger committed:
Remove backend check for torch.compile (#22140)

* Remove backend enforcment for torch.compile

* Update error

* Update src/transformers/training_args.py

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Style

---------

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

patrickvonplaten committed:
[Safetensors] Add explicit  flag to from pretrained (#22083)

* [Safetensors] Add explicit  flag to from pretrained

* add test

* remove @

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

---------

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ydshieh committed:
Prepare daily CI for torch 2.0.0 (#22135)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

MichaelRipa committed:
docs:  New terms and updates to glossary (#21982)

* Updated glossary with new terms, added abbreviations for certain terms and merged autoencoding models, autoregressive models and causal language modeling into encoder and decoder models

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Added link to 'Pipeline for inference' tutorial

* Trigger CI

* Update docs/source/en/glossary.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

* Added entry for self supervised learning, added deleted entries + fixed broken links

* Update docs/source/en/glossary.mdx

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

---------

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
[🛠️] Fix-whisper-breaking-changes  (#21965)

* temp fix

* temporary fix

* update

* fix tests

* fixup

* update based on reveiew

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

* update to fix tests

* update docstring

---------

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

ydshieh committed:
Move `is_pipeline_test_to_skip` to specific model test classes (#21999)

* Move `is_pipeline_test_to_skip` to specific model test classes

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

alaradirik committed:
Add ConvNeXT V2 (#21679)

* Add ConvNeXt V2 to transformers
* TF model is separated from the PR to fix issues

ydshieh committed:
Update 2 doctest expected values for torch 2.0.0 (#22148)

update values

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

nickprock committed:
Translation Italian: perf_train_cpu and perf_train_cpu_many (#22151)

* added translated files

added perf_train_cpu and perf_train_cpu_many

* updated toctree

sgugger committed:
Fix big model inference for T5 models in float16 (#22095)

* Fix big model inference for T5 models in float16

* Apply suggestions from code review

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

* Style

* Trigger CI with latest release

---------

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

alaradirik committed:
Create MaskedImageCompletionOutput and fix ViT docs (#22152)

* create MaskedImageCompletionOutput

* fix bugs

* fix bugs

amyeroberts committed:
to_pil - don't rescale if int and in range 0-255 (#22158)

* Don't rescale if in and in range 0-255

* Raise value error if int values too large

* Update tests/test_image_transforms.py

* Update tests/test_image_transforms.py

stas00 committed:
[trainer] add `--optim adamw_torch_fused` for pt-2.0+ (#22144)

* [trainer] add --optim adamw_torch_fused

* change optim default

* deal with non-torch

* revert default change; prep; add fp16/amp assert

* typo

* typo

sgugger committed:
Revert "Enforce same behavior as PyTorch 2.0 for older versions" (#22163)

Revert "Enforce same behavior as PyTorch 2.0 for older versions (#22136)"

This reverts commit 1c801d65eb42a71ea52db797af760bd96c8b113f.

sgugger committed:
v4.28.0.dev0

sgugger committed:
Load optimizer state on CPU to avoid CUDA OOM (#22159)

sgugger committed:
Run all tests by default (#22162)

shivalikasingh95 committed:
added video instance seg post processing

shivalikasingh95 committed:
conversion script changes for video ckpt

shivalikasingh95 committed:
modeling file fix video example

shivalikasingh95 committed:
add tests

shivalikasingh95 committed:
conversion script minor fix

shivalikasingh95 committed:
Merge branch 'huggingface:main' into add-mask2former-video
